ktor-network/jvm/src/io/ktor/network/sockets/DatagramSocketImpl.kt

```diff
@@ -10,30 +10,26 @@ import io.ktor.util.network.*
 import kotlinx.coroutines.*
 import kotlinx.coroutines.channels.*
 import io.ktor.utils.io.core.*
-import java.net.*
 import java.nio.*
 import java.nio.channels.*
 
-@Suppress("BlockingMethodInNonBlockingContext")
-internal class DatagramSocketImpl(override val channel: DatagramChannel, selector: SelectorManager)
-    : BoundDatagramSocket, ConnectedDatagramSocket, NIOSocketImpl<DatagramChannel>(channel, selector, DefaultDatagramByteBufferPool) {
-
+internal class DatagramSocketImpl(
+    override val channel: DatagramChannel,
+    selector: SelectorManager
+) : BoundDatagramSocket, ConnectedDatagramSocket, NIOSocketImpl<DatagramChannel>(
+    channel, selector, DefaultDatagramByteBufferPool
+) {
     private val socket = channel.socket()!!
 
     override val localAddress: NetworkAddress
-        get() = socket.localSocketAddress as? NetworkAddress
+        get() = socket.localSocketAddress
             ?: throw IllegalStateException("Channel is not yet bound")
 
     override val remoteAddress: NetworkAddress
-        get() = socket.remoteSocketAddress as? NetworkAddress
+        get() = socket.remoteSocketAddress
             ?: throw IllegalStateException("Channel is not yet connected")
 
-    @OptIn(ObsoleteCoroutinesApi::class, ExperimentalCoroutinesApi::class)
-    private val sender = actor<Datagram>(Dispatchers.IO) {
```

Comment:
It appears that because of this pull request, datagram messages are no longer sent on the IO dispatcher but on the dispatcher that it is called from by the user. So for example on Android, if you call `send` on the datagram socket, it will crash with `android.os.NetworkOnMainThreadException`. The solution for the user is to call send from `Dispatchers.IO`, but that is a change in behaviour. Previously, as you can see in the code above, the actor would already use `Dispatchers.IO`. That is now no longer the case. Are you aware of this change in behaviour?

This could be fixed by using `withContext(Dispatchers.IO)` in `DatagramSendChannel.send`. However, the `offer` function would still fail because it cannot suspend. Any idea how to fix this? Should this be fixed? Maybe the `offer` function should not be available? What do you think?

Reply:
`offer` should throw `NetworkOnMainThreadException` to force the user to switch dispatcher manually. Other case we can fix using dispatcher from the socket context

Reply:
Thanks for the investigation, I will add fixes

Reply:
Done, will merge after CI build

Reply:
Great, looks good!